include/aws/s3/private/s3_request.h

```diff
@@ -23,6 +24,69 @@ enum aws_s3_request_flags {
     AWS_S3_REQUEST_FLAG_ALWAYS_SEND = 0x00000004,
 };
 
+/**
+ * Information sent in the meta_request telemetry callback.
+ */
+struct aws_s3_request_metrics {
+    struct aws_allocator *allocator;
+
+    struct {
+        /* The time stamp when the request started by S3 client, which is prepared time by the client. Timestamps
+         * are from `aws_high_res_clock_get_ticks`. This will always be available. */
+        uint64_t start_timestamp_ns;
+        /* The time stamp when the request finished by S3 client succeed or failed or to be retried. Timestamps
+         * are from `aws_high_res_clock_get_ticks`. This will always be available. */
+        uint64_t end_timestamp_ns;
+        /* The time duration for the request from start to finish. end_timestamp_ns - start_timestamp_ns. This will
+         * always be available. */
+        uint64_t total_duration_ns;
+
+        /* The time stamp when the request started to be encoded. 0 means data not available. Timestamp
+         * are from `aws_high_res_clock_get_ticks` */
+        uint64_t send_start_timestamp_ns;
+        /* The time stamp when the request finished to be encoded. 0 means data not available.
+         * Timestamp are from `aws_high_res_clock_get_ticks` */
+        uint64_t send_end_timestamp_ns;
+        /* The time duration for the request from start encoding to finish encoding (send_end_timestamp_ns -
+         * send_start_timestamp_ns). When send_end_timestamp_ns is 0, means data not available. */
+        uint64_t sending_duration_ns;
+
+        /* The time stamp when the response started to be received from the network channel. 0 means data not
+         * available. Timestamp are from `aws_high_res_clock_get_ticks` */
+        uint64_t receive_start_timestamp_ns;
+        /* The time stamp when the response finished to be received from the network channel. 0 means data not
+         * available. Timestamp are from `aws_high_res_clock_get_ticks` */
+        uint64_t receive_end_timestamp_ns;
+        /* The time duration for the request from start receiving to finish receiving. receive_end_timestamp_ns -
+         * receive_start_timestamp_ns. When receive_end_timestamp_ns is 0, means data not available. */
+        uint64_t receiving_duration_ns;
+    } time_metrics;
+
+    struct {
+        /* Response status code for the request */
+        int response_status;
+        /* HTTP Headers of the response received. */
+        struct aws_http_headers *response_headers;
+        /* The request line: method, path and query of the request. */
+        struct aws_byte_buf request_path_query;
+    } req_resp_info_metrics;
+
+    struct {
+        /* The IP address of the request connected to */
+        struct aws_byte_buf ip_address;
+        /* The pointer to the connection that request was made from */
+        void *connection_id;
```

Comment:
exposing void * is just leaking implementation detail that id is just pointer. sjold we do size_t for it instead or something else?

Reply:
changed the getter to return a size_t